### PR TITLE
Adds IItemHandler. A simple, flexible, light-weight, non-terrible version of IInventory/ISidedInventory.

### PR DESCRIPTION
**Pull request by rwtema**
Also adds patches to implement this interface on all vanilla inventory TEs, and to make hoppers and droppers support the interface.

I've also added a SlotItemHandler implementation for GUIs (regular Slot requires an IInventory).

It still needs documentation and possibly a generic implementation, however I'm opening this for feedback/discussion.

### CODE REVIEW

**Comment by rwtema** [comment]
Unlike IInventory/ISidedInventory, all methods include side as a parameter. Because of this, "slot 0" on one side does not have to be "slot 0" on all sides. If needed you can transform the incoming slot parameter to be whatever you need it to be on that particular side (so you could just do "slot = slotsSides[slot]", or whatever, and carry on).

The reason I did it this way is to prevent a "128 slot inventory with no sidedness" from having a large redundant 128 int array of just {1, 2, 3, 4, ..., 127, 128}. That's an incredible waste of memory.

This isn't a perfect approach, and I'm sure some things will be more difficult (especially when dealing with simultaneous querying/insertion/extraction on multiple sides). However for the majority of regular cases, just knowing how many slots are available on a side is more than sufficient.

**Comment by MaPePeR** [comment]
`getStackInSlot` and `extractItem` both return an `ItemStack`. I see upcoming issues where an inventory might return a different `ItemStack` for extraction than what was returned for `getStackInSlot` and the caller is not checking what was actually extracted: Maybe add the 'what i'm trying to extract' as a parameter, so the inventories can easily verify that they talk about the same thing.

Or even worse: The only way an Inventory can disallow extraction from a slot is to return `null` in `extract`: Always need to check that `extractItem` return value...

What is `simulate` supposed to do? If i tell a method that it should not do anything i might as well not call it at all, right? And since there is no guarantee, that someone is going to use the simulate-option in the way my inventory expects it, i can only discard that simulate-information and my `extract` becomes an alias for `getStackInSlot`.

What is `extractItem` supposed to do when the requested amount is bigger than the available amount? return `null` or an `ItemStack` that contains less items? or throw an exception? 
I think the interface should be designed in a way that only one implementation is possible/viable.

Interesting naming: its get*Stack*InSlot but insert and extract*Item*. (Not an issue, only something i just noticed and found funny/interesting)

File: src/main/java/net/minecraftforge/items/ItemHandlerHelper.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		if(dest == null || stack == null) return stack;
+
+		for (int i = 0; i < dest.getSlots(side) && stack != null; i++) {
+			stack = dest.insertItem(i, stack, side, simulate);
+		}
+
+		return stack;
+	}
+
+	public static boolean canItemsStack(ItemStack a, ItemStack b) {
```
Comment (liach) [inline src/main/java/net/minecraftforge/items/ItemHandlerHelper.java]: Bad code style for this whole file - You used tabs and your brackets!

File: src/main/java/net/minecraftforge/items/VanillaInventoryCodeHooks.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.tileentity.TileEntityDispenser;
+import net.minecraft.tileentity.TileEntityHopper;
+import net.minecraft.util.BlockPos;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.world.World;
+import net.minecraftforge.items.wrapper.InvWrapper;
+import net.minecraftforge.items.wrapper.SidedInvWrapper;
+
+public class VanillaInventoryCodeHooks {
```
Comment (liach) [inline src/main/java/net/minecraftforge/items/VanillaInventoryCodeHooks.java]: Bracket

**Comment by liach** [comment]
The code style is too ugly. You totally ignored the code style acquired for forge code. Moreover, your code style is mixed, which showed that you copied some code from mcp (You should not always copy those code)

**Comment by cpw** [comment]
OK @rwtema this is kinda almost what I'm looking for

**Comment by grondag** [comment]
@MaPePeR: I believe the simulate flag that you denigrated is the answer to your concern about inconsistency. 

@rwtema: Want this. Thank you.

**Comment by mezz** [comment]
*drool*
I don't see any glaring issues, and I think Forestry's code would benefit.
I'm sold on the slot count concept, it was my only concern at first.

Maybe deprecate ISidedInventory? For now we can use the wrappers to interact with ISided from other mods, but it would be nice to push adoption along.

I would like to hear thoughts from more modders with item pipes, 
@asiekierka @tterrag1098 @theZorro266

**Comment by asiekierka** [comment]
It's nice, but I want to see what @cpw came up with. This is too important to just push the first solution verbatim.

**Comment by RainWarrior** [comment]
I feel like keeping direct access to underlying ItemStacks is a bad idea - they're still mutable, and people can do whatever they want to them, unless there's defensive copying everywhere (and I don't see it in getStackInSlot, for example, unless that responsibility is delegated to IInventory/ISidedInventory).

**Comment by cpw** [comment]
@asie I would like to use something like this model, but perhaps via the new mechanism I'm putting into tile entities..

**Comment by asiekierka** [comment]
@cpw Using the new mechanism is the only way I ever expected it being
implemented.

@Fry Conversely, creating entire copies of ItemStacks could easily get very
intensive due to NBT copying.

**Comment by RainWarrior** [comment]
One alternative is to use IoC - pass the handler to the inventory/Slot, and only allow the stack modification there; something like
```Java
interface ItemStackHandler()
{
    Tuple4<Item, Integer, Integer, NBTTagCompound> handleStack(Item item, int stackSize, int damage, NBTTagCompound nbt);
}
```
Or, since we're trying to avoid memory allocation:
```Java
interface ItemStackHandler()
{
    void setup(Item item, int stackSize, int damage, NBTTagCompound nbt);
    Item newItem();
    int newStackSize();
    NBTTagCompound newNbt();
}
```
But in this case handler needs to be stateful.

**Comment by asiekierka** [comment]
I say let's not overcomplicate it. This could end poorly.

2015-12-08 21:36 GMT+01:00 Fry <notifications@github.com>:

**Comment by cpw** [comment]
@asie you're right, but at the same time, defensive copying is already spamming a lot of garbage, and a lot of it risks escaping to oldgen where it can cause a serious performance issue.

A function taking an itemstack might be a way to handle it..

**Comment by mezz** [comment]
Time for `ImmutableItemStack`?

**Comment by MaPePeR** [comment]
Isn't creating an immutable ItemStack the same as copying with the additional downside that the receiver needs to copy it again to make use of it?

I think trust has to start *somewhere*. If someone wants to do evil stuff there is no way to stop him/her anyway. 
This can be solved by proper documentation. 

I think I understand the purpose of the simulate option now, but I'm not really happy with it. It's better than what we have right now for sure, though.

**Comment by asiekierka** [comment]
>If someone wants to do evil stuff there is no way to stop him/her anyway. 

Exactly. We need to find a balance between safety and performance - after all, this is a game, and inventory interaction is involved in a huge amount of modding.

**Comment by rwtema** [comment]
Redundant. 
https://github.com/MinecraftForge/MinecraftForge/commit/a1e41f5464e77e726b4ff082aed1f9f117d55832

**Pull request closed without merging**